// geostyler/dist/Component/Renderer/OlRenderer/OlRenderer.js
"use strict";
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OlRenderer = void 0;
var react_1 = __importStar(require("react"));
var Map_1 = __importDefault(require("ol/Map"));
var Vector_1 = __importDefault(require("ol/layer/Vector"));
var Vector_2 = __importDefault(require("ol/source/Vector"));
var Point_1 = __importDefault(require("ol/geom/Point"));
var LineString_1 = __importDefault(require("ol/geom/LineString"));
var Polygon_1 = __importDefault(require("ol/geom/Polygon"));
var Feature_1 = __importDefault(require("ol/Feature"));
var View_1 = __importDefault(require("ol/View"));
var geostyler_openlayers_parser_1 = __importDefault(require("geostyler-openlayers-parser"));
require("./OlRenderer.css");
require("ol/ol.css");
var get_1 = __importDefault(require("lodash/get"));
var uniqueId_1 = __importDefault(require("lodash/uniqueId"));
/**
 * Symbolizer Renderer UI.
 */
var OlRenderer = function (_a) {
    var data = _a.data, onClick = _a.onClick, symbolizerKind = _a.symbolizerKind, symbolizers = _a.symbolizers;
    /** reference to the underlying OpenLayers map */
    var map = (0, react_1.useRef)();
    var layer = (0, react_1.useRef)();
    var mapId = (0, react_1.useState)((0, uniqueId_1.default)('map_'))[0];
    var getSampleGeomFromSymbolizer = (0, react_1.useCallback)(function () {
        var kind = symbolizerKind || (0, get_1.default)(symbolizers, '[0].kind');
        switch (kind) {
            case 'Mark':
            case 'Icon':
            case 'Text':
                return new Point_1.default([7.10066, 50.735851]);
            case 'Fill':
                return new Polygon_1.default([[
                        [7.1031761169433585, 50.734268655851345],
                        [7.109270095825195, 50.734268655851345,],
                        [7.109270095825195, 50.73824770380063],
                        [7.1031761169433585, 50.73824770380063],
                        [7.1031761169433585, 50.734268655851345,]
                    ]]);
            case 'Line':
                return new LineString_1.default([
                    [7.062578201293945, 50.721786104206004],
                    [7.077512741088867, 50.729610159968296],
                    [7.082319259643555, 50.732435192351126],
                    [7.097940444946289, 50.73748722929948],
                    [7.106866836547852, 50.73775882875318],
                    [7.117509841918945, 50.73889952925885],
                    [7.129182815551758, 50.7504679214779]
                ]);
            default:
                return new Point_1.default([7.10066, 50.735851]);
        }
    }, [symbolizerKind, symbolizers]);
    var updateFeature = (0, react_1.useCallback)(function () {
        var exampleFeatureProps = (0, get_1.default)(data, 'exampleFeatures.features[0].properties');
        layer.current.getSource().clear();
        var sampleFeature = new Feature_1.default(__assign({ geometry: getSampleGeomFromSymbolizer(), Name: 'Sample Feature' }, exampleFeatureProps));
        layer.current.getSource().addFeature(sampleFeature);
        // zoom to feature extent
        var extent = layer.current.getSource().getExtent();
        map.current.getView().fit(extent, {
            maxZoom: 20
        });
    }, [data, getSampleGeomFromSymbolizer]);
    (0, react_1.useEffect)(function () {
        layer.current = new Vector_1.default({
            source: new Vector_2.default()
        });
        map.current = new Map_1.default({
            layers: [layer.current],
            controls: [],
            interactions: [],
            target: mapId,
            view: new View_1.default({
                projection: 'EPSG:4326'
            })
        });
    }, [mapId]);
    (0, react_1.useEffect)(function () {
        updateFeature();
    }, [updateFeature]);
    (0, react_1.useEffect)(function () {
        applySymbolizers(symbolizers);
    }, [symbolizers]);
    /**
     * Transforms the incoming symbolizers to an OpenLayers style object the
     * GeoStyler parser and applies it to the vector features on the map.
     *
     * @param {Symbolizer[]} newSymbolizers The symbolizers holding the style to apply
     */
    var applySymbolizers = function (newSymbolizers) { return __awaiter(void 0, void 0, void 0, function () {
        var styleParser, style, _a, olStyles, _b, errors;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    if (!newSymbolizers) {
                        return [2 /*return*/, undefined];
                    }
                    styleParser = new geostyler_openlayers_parser_1.default();
                    style = {
                        name: 'WrapperStyle4Symbolizer',
                        rules: [{
                                name: 'WrapperRule4Symbolizer',
                                symbolizers: newSymbolizers
                            }]
                    };
                    return [4 /*yield*/, styleParser.writeStyle(style)];
                case 1:
                    _a = _c.sent(), olStyles = _a.output, _b = _a.errors, errors = _b === void 0 ? [] : _b;
                    if (errors.length > 0) {
                        return [2 /*return*/, undefined];
                    }
                    else {
                        // apply new OL style to vector layer
                        layer.current.setStyle(olStyles);
                        return [2 /*return*/, olStyles];
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    return (react_1.default.createElement("div", { onClick: function (event) {
            if (onClick) {
                onClick(symbolizers, event);
            }
        }, className: "gs-symbolizer-olrenderer", id: mapId }));
};
exports.OlRenderer = OlRenderer;
exports.default = exports.OlRenderer;
