// geostyler/dist/Component/Symbolizer/MarkEditor/MarkEditor.js
"use strict";
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkEditor = void 0;
var react_1 = __importStar(require("react"));
var WellKnownNameField_1 = __importDefault(require("../Field/WellKnownNameField/WellKnownNameField"));
var WellKnownNameEditor_1 = __importDefault(require("../WellKnownNameEditor/WellKnownNameEditor"));
var LocaleWrapper_1 = require("../../LocaleWrapper/LocaleWrapper");
var en_US_1 = __importDefault(require("../../../locale/en_US"));
var CompositionContext_1 = require("../../../context/CompositionContext/CompositionContext");
var CompositionUtil_1 = __importDefault(require("../../../Util/CompositionUtil"));
var withDefaultsContext_1 = __importDefault(require("../../../hoc/withDefaultsContext"));
var antd_1 = require("antd");
var cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
var UnsupportedPropertiesContext_1 = require("../../../context/UnsupportedPropertiesContext/UnsupportedPropertiesContext");
var UnsupportedPropertiesUtil_1 = __importDefault(require("../../../Util/UnsupportedPropertiesUtil"));
var COMPONENTNAME = 'MarkEditor';
var MarkEditor = function (_a) {
    var _b = _a.locale, locale = _b === void 0 ? en_US_1.default.MarkEditor : _b, symbolizer = _a.symbolizer, onSymbolizerChange = _a.onSymbolizerChange, defaultValues = _a.defaultValues;
    var _c = (0, react_1.useContext)(UnsupportedPropertiesContext_1.UnsupportedPropertiesContext), unsupportedProperties = _c.unsupportedProperties, options = _c.options;
    var onWellKnownNameChange = function (wellKnownName) {
        var clonedSymbolizer = (0, cloneDeep_1.default)(symbolizer);
        clonedSymbolizer.wellKnownName = wellKnownName;
        if (onSymbolizerChange) {
            onSymbolizerChange(clonedSymbolizer);
        }
    };
    var getSupportProps = function (propName) {
        return UnsupportedPropertiesUtil_1.default.getSupportProps(__assign({ propName: propName, symbolizerName: 'MarkSymbolizer', unsupportedProperties: unsupportedProperties }, options));
    };
    return (react_1.default.createElement(CompositionContext_1.CompositionContext.Consumer, null, function (composition) {
        var _a;
        return (react_1.default.createElement("div", { className: "gs-mark-symbolizer-editor" },
            react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.wellKnownNameFieldLabel }, getSupportProps('wellKnownName')), CompositionUtil_1.default.handleComposition({
                composition: composition,
                path: 'MarkEditor.wellKnownNameField',
                onChange: onWellKnownNameChange,
                propName: 'wellKnownName',
                propValue: symbolizer.wellKnownName,
                defaultValue: (_a = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.MarkEditor) === null || _a === void 0 ? void 0 : _a.defaultWellKnownName,
                defaultElement: react_1.default.createElement(WellKnownNameField_1.default, null)
            })),
            react_1.default.createElement(WellKnownNameEditor_1.default, { symbolizer: symbolizer, onSymbolizerChange: onSymbolizerChange })));
    }));
};
exports.MarkEditor = MarkEditor;
exports.default = (0, withDefaultsContext_1.default)((0, LocaleWrapper_1.localize)(exports.MarkEditor, COMPONENTNAME));
