// geostyler/dist/Component/Filter/FilterEditorWindow/FilterEditorWindow.js
"use strict";
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterEditorWindow = void 0;
var React = __importStar(require("react"));
var ReactDOM = __importStar(require("react-dom"));
var react_rnd_1 = require("react-rnd");
require("./FilterEditorWindow.css");
var antd_1 = require("antd");
var icons_1 = require("@ant-design/icons");
var LocaleWrapper_1 = require("../../LocaleWrapper/LocaleWrapper");
var FilterTree_1 = __importDefault(require("../FilterTree/FilterTree"));
var en_US_1 = __importDefault(require("../../../locale/en_US"));
/**
 * Filter Editor Window UI.
 */
var FilterEditorWindow = function (_a) {
    var x = _a.x, y = _a.y, internalDataDef = _a.internalDataDef, onClose = _a.onClose, filter = _a.filter, onFilterChange = _a.onFilterChange, filterUiProps = _a.filterUiProps, _b = _a.locale, locale = _b === void 0 ? en_US_1.default.FilterEditorWindow : _b;
    return (ReactDOM.createPortal(React.createElement(react_rnd_1.Rnd, { className: "filter-editor-window", default: {
            x: x || window.innerWidth / 2,
            y: y || window.innerHeight / 2,
            width: undefined,
            height: undefined
        }, enableResizing: {
            bottom: false,
            bottomLeft: false,
            bottomRight: false,
            left: false,
            right: false,
            top: false,
            topLeft: false,
            topRight: false
        }, dragHandleClassName: "title" },
        React.createElement("div", { className: "header filter-editor-window-header" },
            React.createElement("span", { className: "title" }, locale.filterEditor),
            React.createElement(antd_1.Button, { icon: React.createElement(icons_1.CloseOutlined, null), size: "small", onClick: onClose })),
        React.createElement("div", { className: "filter-editor-window-body" },
            React.createElement(FilterTree_1.default, { internalDataDef: internalDataDef, filter: filter, onFilterChange: onFilterChange, filterUiProps: filterUiProps }))), document.body));
};
exports.FilterEditorWindow = FilterEditorWindow;
exports.default = (0, LocaleWrapper_1.localize)(exports.FilterEditorWindow, 'FilterEditorWindow');
