// geostyler-geojson-parser/build/dist/GeoJsonDataParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 *
 */
var GeoJsonDataParser = /** @class */ (function () {
    function GeoJsonDataParser(sourceProj, targetProj) {
        this.title = 'GeoJSON Data Parser';
        if (sourceProj && targetProj) {
            this.sourceProjection = sourceProj;
            this.targetProjection = targetProj;
        }
    }
    /**
     *
     * @param inputData
     */
    GeoJsonDataParser.prototype.readData = function (inputData) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            try {
                var featureCollection = inputData;
                var schema = _this.parseSchema(featureCollection);
                var data = { schema: schema, exampleFeatures: featureCollection };
                // If we have a valid data object we can bind it to the promise resolver
                resolve(data);
            }
            catch (e) {
                reject(e);
            }
        });
    };
    /**
     *
     * @param geojson {FeatureCollection<GeometryObject>}
     */
    GeoJsonDataParser.prototype.parseSchema = function (geojson) {
        var dataProperties = {};
        var numValues = {};
        var features = geojson.features;
        if (!features || !Array.isArray(features)) {
            throw new Error('Given GeoJSON FeatureCollection does not have a "features" array - EXIT!');
        }
        features.forEach(function (feature) {
            var props = feature.properties;
            if (props) {
                for (var _i = 0, _a = Object.keys(props); _i < _a.length; _i++) {
                    var key = _a[_i];
                    var propVal = props[key];
                    var propType = typeof (propVal);
                    if (Array.isArray(propVal)) {
                        propType = 'array';
                    }
                    if (propType === 'number') {
                        if (!numValues[key]) {
                            numValues[key] = [];
                        }
                        numValues[key].push(propVal);
                    }
                    if (!dataProperties[key]) {
                        dataProperties[key] = {
                            type: propType
                        };
                    }
                }
            }
        });
        for (var _i = 0, _a = Object.keys(numValues); _i < _a.length; _i++) {
            var propKey = _a[_i];
            var nums = numValues[propKey];
            var sortedNums = nums.sort(function (n1, n2) { return n1 - n2; });
            dataProperties[propKey].minimum = sortedNums[0];
            dataProperties[propKey].maximum = sortedNums[sortedNums.length - 1];
        }
        var dataSchema = {
            type: 'object',
            properties: dataProperties
        };
        return dataSchema;
    };
    /**
     * The name of the GeoJsonDataParser.
     */
    GeoJsonDataParser.title = 'GeoJSON Data Parser';
    return GeoJsonDataParser;
}());
exports.GeoJsonDataParser = GeoJsonDataParser;
exports.default = GeoJsonDataParser;
