// geostyler/dist/Component/RuleGenerator/RuleGeneratorWindow.js
"use strict";
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleGeneratorWindow = void 0;
var React = __importStar(require("react"));
var ReactDOM = __importStar(require("react-dom"));
var chroma_js_1 = require("chroma-js");
var react_rnd_1 = require("react-rnd");
require("./RuleGeneratorWindow.css");
var antd_1 = require("antd");
var icons_1 = require("@ant-design/icons");
var LocaleWrapper_1 = require("../LocaleWrapper/LocaleWrapper");
var RuleGenerator_1 = __importDefault(require("./RuleGenerator"));
var isFinite_1 = __importDefault(require("lodash/isFinite"));
var en_US_1 = __importDefault(require("../../locale/en_US"));
var COMPONENTNAME = 'RuleGeneratorWindow';
/**
 * Symbolizer editorwindow UI.
 */
var RuleGeneratorWindow = function (_a) {
    var _b = _a.locale, locale = _b === void 0 ? en_US_1.default.RuleGeneratorWindow : _b, internalDataDef = _a.internalDataDef, x = _a.x, y = _a.y, onClose = _a.onClose, onRulesChange = _a.onRulesChange, colorRamps = _a.colorRamps, useBrewerColorRamps = _a.useBrewerColorRamps, colorSpaces = _a.colorSpaces;
    var ramps = colorRamps;
    if (colorRamps && useBrewerColorRamps) {
        ramps = Object.assign(colorRamps, chroma_js_1.brewer);
    }
    return (ReactDOM.createPortal(React.createElement(react_rnd_1.Rnd, { className: "rule-generator-window", default: {
            x: (0, isFinite_1.default)(x) ? x : window.innerWidth / 2,
            y: (0, isFinite_1.default)(y) ? y : window.innerHeight / 2,
            width: undefined,
            height: undefined
        }, enableResizing: {
            bottom: false,
            bottomLeft: false,
            bottomRight: false,
            left: false,
            right: false,
            top: false,
            topLeft: false,
            topRight: false
        }, bounds: "window", dragHandleClassName: "title" },
        React.createElement("div", { className: "header rule-generator-window-header" },
            React.createElement("span", { className: "title" }, locale.ruleGenerator),
            React.createElement(antd_1.Button, { icon: React.createElement(icons_1.CloseOutlined, null), size: "small", onClick: onClose })),
        React.createElement(RuleGenerator_1.default, { internalDataDef: internalDataDef, onRulesChange: onRulesChange, colorRamps: ramps, colorSpaces: colorSpaces })), document.body));
};
exports.RuleGeneratorWindow = RuleGeneratorWindow;
exports.default = (0, LocaleWrapper_1.localize)(exports.RuleGeneratorWindow, COMPONENTNAME);
