// geostyler/dist/Component/Symbolizer/FillEditor/FillEditor.js
"use strict";
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FillEditor = void 0;
var react_1 = __importStar(require("react"));
var antd_1 = require("antd");
var ColorField_1 = __importDefault(require("../Field/ColorField/ColorField"));
var OpacityField_1 = __importDefault(require("../Field/OpacityField/OpacityField"));
var GraphicEditor_1 = __importDefault(require("../GraphicEditor/GraphicEditor"));
var WidthField_1 = __importDefault(require("../Field/WidthField/WidthField"));
var cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
var get_1 = __importDefault(require("lodash/get"));
var LocaleWrapper_1 = require("../../LocaleWrapper/LocaleWrapper");
var en_US_1 = __importDefault(require("../../../locale/en_US"));
var LineDashField_1 = __importDefault(require("../Field/LineDashField/LineDashField"));
var CompositionContext_1 = require("../../../context/CompositionContext/CompositionContext");
var CompositionUtil_1 = __importDefault(require("../../../Util/CompositionUtil"));
var withDefaultsContext_1 = __importDefault(require("../../../hoc/withDefaultsContext"));
var UnsupportedPropertiesContext_1 = require("../../../context/UnsupportedPropertiesContext/UnsupportedPropertiesContext");
var UnsupportedPropertiesUtil_1 = __importDefault(require("../../../Util/UnsupportedPropertiesUtil"));
var Panel = antd_1.Collapse.Panel;
var COMPONENTNAME = 'FillEditor';
var FillEditor = function (_a) {
    var _b = _a.locale, locale = _b === void 0 ? en_US_1.default.FillEditor : _b, symbolizer = _a.symbolizer, onSymbolizerChange = _a.onSymbolizerChange, defaultValues = _a.defaultValues;
    var _c = (0, react_1.useContext)(UnsupportedPropertiesContext_1.UnsupportedPropertiesContext), unsupportedProperties = _c.unsupportedProperties, options = _c.options;
    var onFillColorChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.color = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onFillOpacityChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.fillOpacity = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onOpacityChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.opacity = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onOutlineOpacityChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.outlineOpacity = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onOutlineColorChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.outlineColor = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onOutlineWidthChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.outlineWidth = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onOutlineDasharrayChange = function (value) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.outlineDasharray = value;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var onGraphicChange = function (newGraphicFill) {
        var symbolizerClone = (0, cloneDeep_1.default)(symbolizer);
        symbolizerClone.graphicFill = newGraphicFill;
        if (onSymbolizerChange) {
            onSymbolizerChange(symbolizerClone);
        }
    };
    var color = symbolizer.color, fillOpacity = symbolizer.fillOpacity, outlineColor = symbolizer.outlineColor, graphicFill = symbolizer.graphicFill, outlineWidth = symbolizer.outlineWidth, outlineDasharray = symbolizer.outlineDasharray, opacity = symbolizer.opacity, outlineOpacity = symbolizer.outlineOpacity;
    var getSupportProps = function (propName) {
        return UnsupportedPropertiesUtil_1.default.getSupportProps(__assign({ propName: propName, symbolizerName: 'FillSymbolizer', unsupportedProperties: unsupportedProperties }, options));
    };
    return (react_1.default.createElement(CompositionContext_1.CompositionContext.Consumer, null, function (composition) {
        var _a, _b, _c, _d, _e, _f;
        return (react_1.default.createElement("div", { className: "gs-fill-symbolizer-editor" },
            react_1.default.createElement(antd_1.Collapse, { bordered: false, defaultActiveKey: ['1'] },
                react_1.default.createElement(Panel, { header: "General", key: "1" },
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.fillColorLabel }, getSupportProps('color')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.fillColorField',
                        onChange: onFillColorChange,
                        propName: 'color',
                        propValue: color,
                        defaultValue: (_a = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _a === void 0 ? void 0 : _a.defaultFillColor,
                        defaultElement: react_1.default.createElement(ColorField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.fillOpacityLabel }, getSupportProps('fillOpacity')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.fillOpacityField',
                        onChange: onFillOpacityChange,
                        propName: 'fillOpacity',
                        propValue: fillOpacity,
                        defaultValue: (_b = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _b === void 0 ? void 0 : _b.defaultFillOpacity,
                        defaultElement: react_1.default.createElement(OpacityField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.opacityLabel }, getSupportProps('opacity')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.opacityField',
                        onChange: onOpacityChange,
                        propName: 'opacity',
                        propValue: opacity,
                        defaultValue: (_c = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _c === void 0 ? void 0 : _c.defaultOpacity,
                        defaultElement: react_1.default.createElement(OpacityField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.outlineOpacityLabel }, getSupportProps('outlineOpacity')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.outlineOpacityField',
                        onChange: onOutlineOpacityChange,
                        propName: 'opacity',
                        propValue: outlineOpacity,
                        defaultValue: (_d = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _d === void 0 ? void 0 : _d.defaultOutlineOpacity,
                        defaultElement: react_1.default.createElement(OpacityField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.outlineColorLabel }, getSupportProps('outlineColor')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.outlineColorField',
                        onChange: onOutlineColorChange,
                        propName: 'color',
                        propValue: outlineColor,
                        defaultValue: (_e = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _e === void 0 ? void 0 : _e.defaultOutlineColor,
                        defaultElement: react_1.default.createElement(ColorField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.outlineWidthLabel }, getSupportProps('outlineWidth')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.outlineWidthField',
                        onChange: onOutlineWidthChange,
                        propName: 'width',
                        propValue: outlineWidth,
                        defaultValue: (_f = defaultValues === null || defaultValues === void 0 ? void 0 : defaultValues.FillEditor) === null || _f === void 0 ? void 0 : _f.defaultOutlineWidth,
                        defaultElement: react_1.default.createElement(WidthField_1.default, null)
                    })),
                    react_1.default.createElement(antd_1.Form.Item, __assign({ label: locale.outlineDasharrayLabel }, getSupportProps('outlineDasharray')), CompositionUtil_1.default.handleComposition({
                        composition: composition,
                        path: 'FillEditor.outlineDasharrayField',
                        onChange: onOutlineDasharrayChange,
                        propName: 'dashArray',
                        propValue: outlineDasharray,
                        defaultElement: react_1.default.createElement(LineDashField_1.default, null)
                    }))),
                react_1.default.createElement(Panel, { header: "Graphic Fill", key: "2" }, CompositionUtil_1.default.handleComposition({
                    composition: composition,
                    path: 'FillEditor.graphicEditorField',
                    onChange: onGraphicChange,
                    onChangeName: 'onGraphicChange',
                    propName: 'graphic',
                    propValue: graphicFill,
                    defaultElement: (react_1.default.createElement(GraphicEditor_1.default, { graphicTypeFieldLabel: locale.graphicFillTypeLabel, graphic: graphicFill, graphicType: (0, get_1.default)(graphicFill, 'kind') }))
                })))));
    }));
};
exports.FillEditor = FillEditor;
exports.default = (0, withDefaultsContext_1.default)((0, LocaleWrapper_1.localize)(exports.FillEditor, COMPONENTNAME));
