// geostyler/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkEditor = exports.localize = exports.locale = exports.LineJoinField = exports.LineEditor = exports.LineDashField = exports.LineCapField = exports.KindField = exports.ImageField = exports.IconSelectorWindow = exports.IconSelector = exports.IconEditor = exports.GrayChannelField = exports.GraphicTypeField = exports.GraphicEditor = exports.GeoStylerContext = exports.GammaField = exports.FontPicker = exports.FilterTree = exports.FilterOverview = exports.FilterEditorWindow = exports.FillEditor = exports.FieldSet = exports.FieldContainer = exports.FadeDurationField = exports.ExtendedField = exports.Editor = exports.DefaultValueContext = exports.DataProvider = exports.DataLoader = exports.ContrastField = exports.ContrastEnhancementField = exports.ConfigProvider = exports.CompositionContext = exports.ComparisonFilter = exports.ColorRampCombo = exports.ColorMapTypeField = exports.ColorMapEntryField = exports.ColorMapEditor = exports.ColorField = exports.CodeEditor = exports.ChannelSelectionField = exports.ChannelField = exports.CardStyle = exports.BulkEditor = exports.BulkEditModals = exports.BrightnessField = exports.Breadcrumb = exports.BoolFilterField = exports.AttributeCombo = void 0;
exports.WellKnownNameField = exports.WellKnownNameEditor = exports.UnsupportedPropertiesContext = exports.UploadButton = exports.TextFilterField = exports.TextEditor = exports.Symbolizers = exports.SymbolizerEditorWindow = exports.SymbolizerCard = exports.StyleOverview = exports.StyleLoader = exports.StyleFieldContainer = exports.Style = exports.SourceChannelNameField = exports.SortableItem = exports.SLDRenderer = exports.SizeField = exports.SelectableItem = exports.Selectable = exports.ScaleDenominator = exports.SaturationField = exports.RuleTable = exports.Rules = exports.RuleOverview = exports.RuleGeneratorWindow = exports.RuleGenerator = exports.RuleFieldContainer = exports.RuleCard = exports.Rule = exports.RotateField = exports.ResamplingField = exports.RgbChannelField = exports.Renderer = exports.RemovableItem = exports.RemoveButton = exports.RasterEditor = exports.RasterChannelEditor = exports.RadiusField = exports.PropTextEditor = exports.PreviewMap = exports.Preview = exports.OperatorCombo = exports.OpacityField = exports.OlRenderer = exports.OffsetField = exports.NumberFilterField = exports.NameField = exports.MultiEditor = exports.MinScaleDenominator = exports.MaxScaleDenominator = void 0;
exports.withDefaultsContext = exports.WidthField = void 0;
/* eslint-disable camelcase */
/* Released under the BSD 2-Clause License
 *
 * Copyright © 2018-present, terrestris GmbH & Co. KG and GeoStyler contributors
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.
 *
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
var AttributeCombo_1 = __importDefault(require("./Component/Filter/AttributeCombo/AttributeCombo"));
exports.AttributeCombo = AttributeCombo_1.default;
var BoolFilterField_1 = __importDefault(require("./Component/Filter/BoolFilterField/BoolFilterField"));
exports.BoolFilterField = BoolFilterField_1.default;
var Breadcrumb_1 = __importDefault(require("./Component/Breadcrumb/Breadcrumb"));
exports.Breadcrumb = Breadcrumb_1.default;
var BrightnessField_1 = __importDefault(require("./Component/Symbolizer/Field/BrightnessField/BrightnessField"));
exports.BrightnessField = BrightnessField_1.default;
var BulkEditModals_1 = __importDefault(require("./Component/Symbolizer/BulkEditModals/BulkEditModals"));
exports.BulkEditModals = BulkEditModals_1.default;
var BulkEditor_1 = __importDefault(require("./Component/BulkEditor/BulkEditor"));
exports.BulkEditor = BulkEditor_1.default;
var CardStyle_1 = __importDefault(require("./Component/CardStyle/CardStyle"));
exports.CardStyle = CardStyle_1.default;
var ChannelField_1 = __importDefault(require("./Component/Symbolizer/Field/ChannelField/ChannelField"));
exports.ChannelField = ChannelField_1.default;
var ChannelSelectionField_1 = __importDefault(require("./Component/Symbolizer/Field/ChannelSelectionField/ChannelSelectionField"));
exports.ChannelSelectionField = ChannelSelectionField_1.default;
var CodeEditor_1 = __importDefault(require("./Component/CodeEditor/CodeEditor"));
exports.CodeEditor = CodeEditor_1.default;
var ColorField_1 = __importDefault(require("./Component/Symbolizer/Field/ColorField/ColorField"));
exports.ColorField = ColorField_1.default;
var ColorMapEditor_1 = __importDefault(require("./Component/Symbolizer/ColorMapEditor/ColorMapEditor"));
exports.ColorMapEditor = ColorMapEditor_1.default;
var ColorMapEntryField_1 = __importDefault(require("./Component/Symbolizer/Field/ColorMapEntryField/ColorMapEntryField"));
exports.ColorMapEntryField = ColorMapEntryField_1.default;
var ColorMapTypeField_1 = __importDefault(require("./Component/Symbolizer/Field/ColorMapTypeField/ColorMapTypeField"));
exports.ColorMapTypeField = ColorMapTypeField_1.default;
var ColorRampCombo_1 = __importDefault(require("./Component/RuleGenerator/ColorRampCombo/ColorRampCombo"));
exports.ColorRampCombo = ColorRampCombo_1.default;
var ComparisonFilter_1 = __importDefault(require("./Component/Filter/ComparisonFilter/ComparisonFilter"));
exports.ComparisonFilter = ComparisonFilter_1.default;
var ContrastEnhancementField_1 = __importDefault(require("./Component/Symbolizer/Field/ContrastEnhancementField/ContrastEnhancementField"));
exports.ContrastEnhancementField = ContrastEnhancementField_1.default;
var ContrastField_1 = __importDefault(require("./Component/Symbolizer/Field/ContrastField/ContrastField"));
exports.ContrastField = ContrastField_1.default;
var DataLoader_1 = __importDefault(require("./Component/DataInput/DataLoader/DataLoader"));
exports.DataLoader = DataLoader_1.default;
var DataProvider_1 = __importDefault(require("./DataProvider/DataProvider"));
exports.DataProvider = DataProvider_1.default;
var Editor_1 = __importDefault(require("./Component/Symbolizer/Editor/Editor"));
exports.Editor = Editor_1.default;
var ExtendedField_1 = __importDefault(require("./Component/Symbolizer/Field/ExtendedField/ExtendedField"));
exports.ExtendedField = ExtendedField_1.default;
var FadeDurationField_1 = __importDefault(require("./Component/Symbolizer/Field/FadeDurationField/FadeDurationField"));
exports.FadeDurationField = FadeDurationField_1.default;
var FieldSet_1 = __importDefault(require("./Component/FieldSet/FieldSet"));
exports.FieldSet = FieldSet_1.default;
var FieldContainer_1 = __importDefault(require("./Component/FieldContainer/FieldContainer"));
exports.FieldContainer = FieldContainer_1.default;
var FillEditor_1 = __importDefault(require("./Component/Symbolizer/FillEditor/FillEditor"));
exports.FillEditor = FillEditor_1.default;
var FilterEditorWindow_1 = __importDefault(require("./Component/Filter/FilterEditorWindow/FilterEditorWindow"));
exports.FilterEditorWindow = FilterEditorWindow_1.default;
var FilterOverview_1 = __importDefault(require("./Component/FilterOverview/FilterOverview"));
exports.FilterOverview = FilterOverview_1.default;
var FilterTree_1 = __importDefault(require("./Component/Filter/FilterTree/FilterTree"));
exports.FilterTree = FilterTree_1.default;
var FontPicker_1 = __importDefault(require("./Component/Symbolizer/Field/FontPicker/FontPicker"));
exports.FontPicker = FontPicker_1.default;
var GammaField_1 = __importDefault(require("./Component/Symbolizer/Field/GammaField/GammaField"));
exports.GammaField = GammaField_1.default;
var GraphicEditor_1 = __importDefault(require("./Component/Symbolizer/GraphicEditor/GraphicEditor"));
exports.GraphicEditor = GraphicEditor_1.default;
var GraphicTypeField_1 = __importDefault(require("./Component/Symbolizer/Field/GraphicTypeField/GraphicTypeField"));
exports.GraphicTypeField = GraphicTypeField_1.default;
var GrayChannelField_1 = __importDefault(require("./Component/Symbolizer/Field/GrayChannelField/GrayChannelField"));
exports.GrayChannelField = GrayChannelField_1.default;
var IconEditor_1 = __importDefault(require("./Component/Symbolizer/IconEditor/IconEditor"));
exports.IconEditor = IconEditor_1.default;
var IconSelector_1 = __importDefault(require("./Component/Symbolizer/IconSelector/IconSelector"));
exports.IconSelector = IconSelector_1.default;
var IconSelectorWindow_1 = __importDefault(require("./Component/Symbolizer/IconSelectorWindow/IconSelectorWindow"));
exports.IconSelectorWindow = IconSelectorWindow_1.default;
var ImageField_1 = __importDefault(require("./Component/Symbolizer/Field/ImageField/ImageField"));
exports.ImageField = ImageField_1.default;
var KindField_1 = __importDefault(require("./Component/Symbolizer/Field/KindField/KindField"));
exports.KindField = KindField_1.default;
var LineCapField_1 = __importDefault(require("./Component/Symbolizer/Field/LineCapField/LineCapField"));
exports.LineCapField = LineCapField_1.default;
var LineDashField_1 = __importDefault(require("./Component/Symbolizer/Field/LineDashField/LineDashField"));
exports.LineDashField = LineDashField_1.default;
var LineEditor_1 = __importDefault(require("./Component/Symbolizer/LineEditor/LineEditor"));
exports.LineEditor = LineEditor_1.default;
var LineJoinField_1 = __importDefault(require("./Component/Symbolizer/Field/LineJoinField/LineJoinField"));
exports.LineJoinField = LineJoinField_1.default;
var MarkEditor_1 = __importDefault(require("./Component/Symbolizer/MarkEditor/MarkEditor"));
exports.MarkEditor = MarkEditor_1.default;
var MaxScaleDenominator_1 = __importDefault(require("./Component/ScaleDenominator/MaxScaleDenominator"));
exports.MaxScaleDenominator = MaxScaleDenominator_1.default;
var MinScaleDenominator_1 = __importDefault(require("./Component/ScaleDenominator/MinScaleDenominator"));
exports.MinScaleDenominator = MinScaleDenominator_1.default;
var MultiEditor_1 = __importDefault(require("./Component/Symbolizer/MultiEditor/MultiEditor"));
exports.MultiEditor = MultiEditor_1.default;
var NameField_1 = __importDefault(require("./Component/NameField/NameField"));
exports.NameField = NameField_1.default;
var NumberFilterField_1 = __importDefault(require("./Component/Filter/NumberFilterField/NumberFilterField"));
exports.NumberFilterField = NumberFilterField_1.default;
var OffsetField_1 = __importDefault(require("./Component/Symbolizer/Field/OffsetField/OffsetField"));
exports.OffsetField = OffsetField_1.default;
var OpacityField_1 = __importDefault(require("./Component/Symbolizer/Field/OpacityField/OpacityField"));
exports.OpacityField = OpacityField_1.default;
var OperatorCombo_1 = __importDefault(require("./Component/Filter/OperatorCombo/OperatorCombo"));
exports.OperatorCombo = OperatorCombo_1.default;
var Preview_1 = __importDefault(require("./Component/Symbolizer/Preview/Preview"));
exports.Preview = Preview_1.default;
var PreviewMap_1 = __importDefault(require("./Component/PreviewMap/PreviewMap"));
exports.PreviewMap = PreviewMap_1.default;
var PropTextEditor_1 = __importDefault(require("./Component/Symbolizer/PropTextEditor/PropTextEditor"));
exports.PropTextEditor = PropTextEditor_1.default;
var RadiusField_1 = __importDefault(require("./Component/Symbolizer/Field/RadiusField/RadiusField"));
exports.RadiusField = RadiusField_1.default;
var RasterChannelEditor_1 = __importDefault(require("./Component/Symbolizer/RasterChannelEditor/RasterChannelEditor"));
exports.RasterChannelEditor = RasterChannelEditor_1.default;
var RasterEditor_1 = __importDefault(require("./Component/Symbolizer/RasterEditor/RasterEditor"));
exports.RasterEditor = RasterEditor_1.default;
var RemovableItem_1 = __importDefault(require("./Component/RemovableItem/RemovableItem"));
exports.RemovableItem = RemovableItem_1.default;
var RemoveButton_1 = __importDefault(require("./Component/Rule/RemoveButton/RemoveButton"));
exports.RemoveButton = RemoveButton_1.default;
var OlRenderer_1 = __importDefault(require("./Component/Renderer/OlRenderer/OlRenderer"));
exports.OlRenderer = OlRenderer_1.default;
var Renderer_1 = __importDefault(require("./Component/Renderer/Renderer/Renderer"));
exports.Renderer = Renderer_1.default;
var RgbChannelField_1 = __importDefault(require("./Component/Symbolizer/Field/RgbChannelField/RgbChannelField"));
exports.RgbChannelField = RgbChannelField_1.default;
var ResamplingField_1 = __importDefault(require("./Component/Symbolizer/Field/ResamplingField/ResamplingField"));
exports.ResamplingField = ResamplingField_1.default;
var RotateField_1 = __importDefault(require("./Component/Symbolizer/Field/RotateField/RotateField"));
exports.RotateField = RotateField_1.default;
var Rule_1 = __importDefault(require("./Component/Rule/Rule"));
exports.Rule = Rule_1.default;
var RuleCard_1 = __importDefault(require("./Component/RuleCard/RuleCard"));
exports.RuleCard = RuleCard_1.default;
var RuleFieldContainer_1 = __importDefault(require("./Component/RuleFieldContainer/RuleFieldContainer"));
exports.RuleFieldContainer = RuleFieldContainer_1.default;
var RuleGenerator_1 = __importDefault(require("./Component/RuleGenerator/RuleGenerator"));
exports.RuleGenerator = RuleGenerator_1.default;
var RuleGeneratorWindow_1 = __importDefault(require("./Component/RuleGenerator/RuleGeneratorWindow"));
exports.RuleGeneratorWindow = RuleGeneratorWindow_1.default;
var RuleOverview_1 = __importDefault(require("./Component/RuleOverview/RuleOverview"));
exports.RuleOverview = RuleOverview_1.default;
var Rules_1 = __importDefault(require("./Component/Rules/Rules"));
exports.Rules = Rules_1.default;
var RuleTable_1 = __importDefault(require("./Component/RuleTable/RuleTable"));
exports.RuleTable = RuleTable_1.default;
var SaturationField_1 = __importDefault(require("./Component/Symbolizer/Field/SaturationField/SaturationField"));
exports.SaturationField = SaturationField_1.default;
var ScaleDenominator_1 = __importDefault(require("./Component/ScaleDenominator/ScaleDenominator"));
exports.ScaleDenominator = ScaleDenominator_1.default;
var Selectable_1 = __importDefault(require("./Component/Selectable/Selectable"));
exports.Selectable = Selectable_1.default;
var SelectableItem_1 = __importDefault(require("./Component/Selectable/SelectableItem/SelectableItem"));
exports.SelectableItem = SelectableItem_1.default;
var SizeField_1 = __importDefault(require("./Component/Symbolizer/Field/SizeField/SizeField"));
exports.SizeField = SizeField_1.default;
var SortableItem_1 = __importDefault(require("./Component/SortableItem/SortableItem"));
exports.SortableItem = SortableItem_1.default;
var SLDRenderer_1 = __importDefault(require("./Component/Renderer/SLDRenderer/SLDRenderer"));
exports.SLDRenderer = SLDRenderer_1.default;
var SourceChannelNameField_1 = __importDefault(require("./Component/Symbolizer/Field/SourceChannelNameField/SourceChannelNameField"));
exports.SourceChannelNameField = SourceChannelNameField_1.default;
var Style_1 = __importDefault(require("./Component/Style/Style"));
exports.Style = Style_1.default;
var StyleFieldContainer_1 = __importDefault(require("./Component/StyleFieldContainer/StyleFieldContainer"));
exports.StyleFieldContainer = StyleFieldContainer_1.default;
var StyleLoader_1 = __importDefault(require("./Component/DataInput/StyleLoader/StyleLoader"));
exports.StyleLoader = StyleLoader_1.default;
var StyleOverview_1 = __importDefault(require("./Component/StyleOverview/StyleOverview"));
exports.StyleOverview = StyleOverview_1.default;
var SymbolizerCard_1 = __importDefault(require("./Component/SymbolizerCard/SymbolizerCard"));
exports.SymbolizerCard = SymbolizerCard_1.default;
var SymbolizerEditorWindow_1 = __importDefault(require("./Component/Symbolizer/SymbolizerEditorWindow/SymbolizerEditorWindow"));
exports.SymbolizerEditorWindow = SymbolizerEditorWindow_1.default;
var Symbolizers_1 = __importDefault(require("./Component/Symbolizers/Symbolizers"));
exports.Symbolizers = Symbolizers_1.default;
var TextEditor_1 = __importDefault(require("./Component/Symbolizer/TextEditor/TextEditor"));
exports.TextEditor = TextEditor_1.default;
var TextFilterField_1 = __importDefault(require("./Component/Filter/TextFilterField/TextFilterField"));
exports.TextFilterField = TextFilterField_1.default;
var UploadButton_1 = __importDefault(require("./Component/UploadButton/UploadButton"));
exports.UploadButton = UploadButton_1.default;
var WellKnownNameEditor_1 = __importDefault(require("./Component/Symbolizer/WellKnownNameEditor/WellKnownNameEditor"));
exports.WellKnownNameEditor = WellKnownNameEditor_1.default;
var WellKnownNameField_1 = __importDefault(require("./Component/Symbolizer/Field/WellKnownNameField/WellKnownNameField"));
exports.WellKnownNameField = WellKnownNameField_1.default;
var WidthField_1 = __importDefault(require("./Component/Symbolizer/Field/WidthField/WidthField"));
exports.WidthField = WidthField_1.default;
var withDefaultsContext_1 = __importDefault(require("./hoc/withDefaultsContext"));
exports.withDefaultsContext = withDefaultsContext_1.default;
var LocaleWrapper_1 = require("./Component/LocaleWrapper/LocaleWrapper");
Object.defineProperty(exports, "localize", { enumerable: true, get: function () { return LocaleWrapper_1.localize; } });
var CompositionContext_1 = require("./context/CompositionContext/CompositionContext");
Object.defineProperty(exports, "CompositionContext", { enumerable: true, get: function () { return CompositionContext_1.CompositionContext; } });
var DefaultValueContext_1 = require("./context/DefaultValueContext/DefaultValueContext");
Object.defineProperty(exports, "DefaultValueContext", { enumerable: true, get: function () { return DefaultValueContext_1.DefaultValueContext; } });
var UnsupportedPropertiesContext_1 = require("./context/UnsupportedPropertiesContext/UnsupportedPropertiesContext");
Object.defineProperty(exports, "UnsupportedPropertiesContext", { enumerable: true, get: function () { return UnsupportedPropertiesContext_1.UnsupportedPropertiesContext; } });
var GeoStylerContext_1 = require("./context/GeoStylerContext/GeoStylerContext");
Object.defineProperty(exports, "GeoStylerContext", { enumerable: true, get: function () { return GeoStylerContext_1.GeoStylerContext; } });
var antd_1 = require("antd");
Object.defineProperty(exports, "ConfigProvider", { enumerable: true, get: function () { return antd_1.ConfigProvider; } });
var de_DE_1 = __importDefault(require("./locale/de_DE"));
var en_US_1 = __importDefault(require("./locale/en_US"));
var es_ES_1 = __importDefault(require("./locale/es_ES"));
var fr_FR_1 = __importDefault(require("./locale/fr_FR"));
var zh_CN_1 = __importDefault(require("./locale/zh_CN"));
var locale = {
    de_DE: de_DE_1.default,
    en_US: en_US_1.default,
    es_ES: es_ES_1.default,
    fr_FR: fr_FR_1.default,
    zh_CN: zh_CN_1.default
};
exports.locale = locale;
